temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java

```diff
@@ -913,20 +915,33 @@ private <R> R mutableSideEffectImpl(
   @Override
   public int getVersion(String changeId, int minSupported, int maxSupported) {
     CompletablePromise<Integer> result = Workflow.newPromise();
-    replayContext.getVersion(
-        changeId,
-        minSupported,
-        maxSupported,
-        (v, e) ->
-            runner.executeInWorkflowThread(
-                "version-callback",
-                () -> {
-                  if (v != null) {
-                    result.complete(v);
-                  } else {
-                    result.completeExceptionally(e);
-                  }
-                }));
+    boolean markerExists =
+        replayContext.getVersion(
+            changeId,
+            minSupported,
+            maxSupported,
+            (v, e) ->
+                runner.executeInWorkflowThread(
+                    "version-callback",
+                    () -> {
+                      if (v != null) {
+                        result.complete(v);
+                      } else {
+                        result.completeExceptionally(e);
+                      }
+                    }));
+    /*
+     * If we are replaying a workflow and encounter a getVersion call it is possible that this call did not exist
+     * on the original execution. If the call did not exist on the original execution then we cannot block on results
+     * because it can lead to non-deterministic scheduling.
+     * */
```

Comment:
I am not understanding this exactly. Are you saying that before now, Java would never return default version if `getVersion` was called during replay?

Reply:
No, I am not sure what your confused about

Reply:
Sorry, I think I am a bit dense. So how does this code skip a yield like the flag name says?

I think a test that demonstrates this issue from scratch may be clearest to me. You can just send queries to trigger replay (or stop/start worker w/ new code version). I am not following the test to demonstrate the issue. Even just standalone code here may help if we're not willing to be able to disable flag setting during a test.

Reply:
Sorry wasn't trying to be rude I just wasn't sure what you were confused about so I could explain.

> So how does this code skip a yield like the flag name says?

this if statement is skipping the yield since the function will eagerly return if it succeeds. If it fails it will try to call `result.get();` and that will yield back to the scheduler and introduce non determinism


There is a test that demonstrates the issue `testGetVersionDefaultMultithreading`

Reply:
> Sorry wasn't trying to be rude I just wasn't sure what you were confused about so I could explain.

I totally read it as a simple request for more info, no rudeness detected, no worries

> this if statement is skipping the yield since the function will eagerly return if it succeeds

Ah, I see. What about other cases where get version yields, is that not equally concerning? Or is the yield so predictable that so long as `getVersion` is called, it is predictable regardless of what the values set in it are?

So to confirm, this works a little bit better for you than a SDK flag like `SdkFlag.GET_VERSION_AS_SYNCHRONOUS`? I understand from your issue description that it is complicated. Yeah, I just took a dive in those state machine classes and they are complicated.

Question, can you confirm that by the time you get to this line of code the promise has been set? If not, where in the Java code could it be yielding (not counting the `get` below which, after reading the code, doesn't yield unless the promise is not already completed)?

Reply:
>Ah, I see. What about other cases where get version yields, is that not equally concerning? Or is the yield so predictable that so long as getVersion is called, it is predictable regardless of what the values set in it are?

yes, the issues isn't yielding, it is the fact that on replay it would yield, but not executing for the first time

>So to confirm, this works a little bit better for you than a SDK flag like SdkFlag.GET_VERSION_AS_SYNCHRONOUS? 

yeah, but neither solution is great

>Question, can you confirm/guarantee that by the time you get to this line of code the promise has been set? If not, where in the Java code could it be yielding (not counting the get below which, after reading the code, doesn't yield unless the promise is not already completed)?

No, I am pretty sure I can guarantee that by the time the `get()` is reached the promise is set

Reply:
> No, I am pretty sure I can guarantee that by the time the get() is reached the promise is NOT set

Is that only because `runner.executeInWorkflowThread(` or is there another thing happening inside the code that may be non-synchronous. I wonder what would happen if, with an SDK flag, you just ran the code instead wrapped in `executeInWorkflowThread`.

Reply:
its because the state machine , like almost all state machines in java, changes state based on command events.

Reply:
I see, and between events is where it can yield? If so, then I'll approve this PR as is. I just want to confirm that if you removed the `executeInWorkflowThread` outer part in the callback that the promise isn't resolved synchronously (_before_ you call `get` which does yield to other async things).

Reply:
Yes, well commands specifically.

removing the `executeInWorkflowThread` will also cause another breaking change in behaviour because it will change how the callback is scheduled

Reply:
Right, I only want to test with that. Then only remove it when an SDK flag is set, so the flag would then be more general purpose for making `getVersion` sync. But if you're saying the removal of that call does not help make the promise setting synchronous, then nevermind.

Reply:
Yeah removing it does not make the promise setting synchronous. It also shouldn't be removed because then you could end up running\ workflow code in a non workflow thread which isn't really correct.

Reply:
:+1: Can consider this thread resolved. Will approve as other things are addressed. Thanks for explaining this all to me.